unison-cli/src/Unison/Codebase/Editor/Slurp.hs

```diff
@@ -0,0 +1,497 @@
+
+-- | This semigroup is how a definition's status is determined.
+instance Semigroup (DefStatus v) where
+  -- If the definition has its own error, that takes highest priority.
+  SelfErr err <> _ = SelfErr err
+  _ <> SelfErr err = SelfErr err
+  -- Next we care if a dependency has an error
+  ErrFrom v err <> _ = ErrFrom v err
+  _ <> ErrFrom v err = ErrFrom v err
+  -- If our definition needs its own update then we don't care if dependencies need updates.
+  Ok Updated <> _ = Ok Updated
+  _ <> Ok Updated = Ok Updated
+  NeedsUpdate v <> _ = NeedsUpdate v
+  _ <> NeedsUpdate v = NeedsUpdate v
+  -- 'New' definitions take precedence over duplicated dependencies
+  Ok New <> _ = Ok New
+  _ <> Ok New = Ok New
+  Ok Duplicated <> _ = Ok Duplicated
+
+-- | Possible error conditions for a definition.
+data SlurpErr
+  = TermCtorCollision
+  | CtorTermCollision
+  | Conflict
+  deriving (Eq, Ord, Show)
+
+-- | Possible error conditions for a definition.
+data DefinitionNotes
+  = DefOk SlurpStatus
+  | DefErr SlurpErr
+  deriving (Show)
+
+-- | A map of variables to their status, and all of their dependencies' statuses.
+type SlurpAnalysis v = Map (TermOrTypeVar v) (DefinitionNotes, Map (TermOrTypeVar v) DefinitionNotes)
+
+-- | A mapping of each status to the vars which have that status.
+type VarsByStatus v = Map (DefStatus v) (NESet (TermOrTypeVar v))
+
+-- Compute all definitions which can be added, or the reasons why a def can't be added.
+groupByStatus :: forall v. (Ord v, Show v) => SlurpAnalysis v -> VarsByStatus v
+groupByStatus sr =
+  pTraceShowId $ analyzed
+  where
+    analyzed :: Map (DefStatus v) (NESet (TermOrTypeVar v))
+    analyzed =
+      sr
+        & Map.toList
+        & fmap
+          ( \(tv, (defNotes, deps)) ->
+              ( Semigroup.fold1 (computeDefStatus False defNotes tv NEList.:| (Map.toList deps <&> \(depTV, depDefNotes) -> computeDefStatus True depDefNotes depTV)),
+                NESet.singleton tv
+              )
+          )
+        & Map.fromListWith (<>)
+
+    computeDefStatus :: (Ord v, Show v) => Bool -> DefinitionNotes -> TermOrTypeVar v -> DefStatus v
+    computeDefStatus isDep defNotes tv =
+      case defNotes of
+        DefOk Updated -> if isDep then NeedsUpdate tv else Ok Updated
+        DefErr err -> ErrFrom tv err
+        DefOk New -> Ok New
+        DefOk Duplicated -> Ok Duplicated
+
+-- | Analyze a file and determine the status of all of its definitions with respect to a set
+-- of vars to analyze and an operation you wish to perform.
+slurpFile ::
+  forall v.
+  Var v =>
+  UF.TypecheckedUnisonFile v Ann ->
+  Set v ->
+  Maybe SlurpOp ->
+  Names ->
+  SR.SlurpResult v
+slurpFile uf defsToConsider maybeSlurpOp unalteredCodebaseNames =
+  let varRelation :: Rel.Relation (TermOrTypeVar v) LD.LabeledDependency
+      varRelation = fileDefinitions uf
+      involvedVars :: Set (TermOrTypeVar v)
+      involvedVars = computeInvolvedVars uf defsToConsider varRelation
```

Comment:
`involvedVars` is every defToConsider and all of its in-file transitive dependencies.